File: mns-core/src/test/java/cz/fb/manaus/core/category/CategoryServiceRealDataTest.java
package cz.fb.manaus.core.category;

import com.google.common.collect.HashMultiset;
import com.google.common.collect.Sets;
import cz.fb.manaus.core.MarketCategories;
import cz.fb.manaus.core.category.categorizer.CountryCodeCategorizer;
import cz.fb.manaus.core.category.categorizer.RunnerCountCategorizer;
import cz.fb.manaus.core.category.categorizer.SportCategorizer;
import cz.fb.manaus.core.manager.AbstractMarketDataAwareTestCase;
import org.junit.Test;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.Set;

import static junit.framework.TestCase.assertTrue;
import static org.hamcrest.CoreMatchers.containsString;
import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;

public class CategoryServiceRealDataTest extends AbstractMarketDataAwareTestCase {
    public static final Set<String> DISJUNCTIVE_CATEGORIES = Set.of(
            Category.MARKET_PREFIX + SportCategorizer.PREFIX + MarketCategories.BASKETBALL,
            Category.MARKET_PREFIX + SportCategorizer.PREFIX + MarketCategories.VOLLEYBALL,
            Category.MARKET_PREFIX + SportCategorizer.PREFIX + MarketCategories.BASEBALL,
            Category.MARKET_PREFIX + SportCategorizer.PREFIX + MarketCategories.SOCCER,
            Category.MARKET_PREFIX + SportCategorizer.PREFIX + MarketCategories.AMERICAN_FOOTBALL,
            Category.MARKET_PREFIX + SportCategorizer.PREFIX + MarketCategories.GOLF,
            Category.MARKET_PREFIX + SportCategorizer.PREFIX + MarketCategories.HANDBALL,
            Category.MARKET_PREFIX + SportCategorizer.PREFIX + MarketCategories.GREY_HOUNDS,
            Category.MARKET_PREFIX + SportCategorizer.PREFIX + MarketCategories.TENNIS,
            Category.MARKET_PREFIX + SportCategorizer.PREFIX + MarketCategories.FINANCIAL,
            Category.MARKET_PREFIX + SportCategorizer.PREFIX + MarketCategories.SNOOKER,
            Category.MARKET_PREFIX + SportCategorizer.PREFIX + MarketCategories.CRICKET,
            Category.MARKET_PREFIX + SportCategorizer.PREFIX + MarketCategories.RUGBY,
            Category.MARKET_PREFIX + SportCategorizer.PREFIX + MarketCategories.ICE_HOCKEY,
            Category.MARKET_PREFIX + SportCategorizer.PREFIX + MarketCategories.MOTOR_SPORT,
            Category.MARKET_PREFIX + SportCategorizer.PREFIX + MarketCategories.HORSES);

    @Autowired
    private CategoryService categoryService;

    @Test
    public void testDisjunctiveCategories() {
        for (var market : markets) {
            var marketCategories = categoryService.getMarketCategories(market, false);
            var intersection = Sets.intersection(marketCategories, DISJUNCTIVE_CATEGORIES);
            assertThat(market.toString(), intersection.size(), is(1));
        }
    }

    @Test
    public void testSoccer() {
        checkCategory(Category.MARKET_PREFIX + SportCategorizer.PREFIX + MarketCategories.SOCCER, "soccer");
    }

    @Test
    public void testBasket() {
        checkCategory(Category.MARKET_PREFIX + SportCategorizer.PREFIX + MarketCategories.BASKETBALL, "basket");
    }

    @Test
    public void testTennis() {
        checkCategory(Category.MARKET_PREFIX + SportCategorizer.PREFIX + MarketCategories.TENNIS, "tennis");
    }

    @Test
    public void testAmericanFootball() {
        checkCategory(Category.MARKET_PREFIX + SportCategorizer.PREFIX + MarketCategories.AMERICAN_FOOTBALL, "american");
    }

    @Test
    public void testIceHockey() {
        checkCategory(Category.MARKET_PREFIX + SportCategorizer.PREFIX + MarketCategories.ICE_HOCKEY, "hockey");
    }

    @Test
    public void testGreyhounds() {
        checkCategory(Category.MARKET_PREFIX + SportCategorizer.PREFIX + MarketCategories.GREY_HOUNDS, "hound");
    }

    @Test
    public void testVolleyball() {
        checkCategory(Category.MARKET_PREFIX + SportCategorizer.PREFIX + MarketCategories.VOLLEYBALL, "volley");
    }

    @Test
    public void testGolf() {
        checkCategory(Category.MARKET_PREFIX + SportCategorizer.PREFIX + MarketCategories.GOLF, "golf");
    }

    @Test
    public void testRunnerCount() {
        checkCategory(Category.MARKET_PREFIX + RunnerCountCategorizer.PREFIX + "4", null);
    }

    @Test
    public void testUkraine() {
        checkCategory(Category.MARKET_PREFIX + CountryCodeCategorizer.PREFIX + "ua", null);
    }

    @Test
    public void testFinancialBets() {
        checkCategory(Category.MARKET_PREFIX + SportCategorizer.PREFIX + MarketCategories.FINANCIAL, "financial");
    }

    private void checkCategory(String category, String mustContainLower) {
        var size = markets.size();
        var count = getCategoryCount(category, mustContainLower);
        assertTrue(count > 0 && count < size);
    }

    private int getCategoryCount(String category, String mustContainLower) {
        var counts = HashMultiset.<String>create();
        for (var market : markets) {
            var categories = categoryService.getMarketCategories(market, false);
            counts.addAll(categories);
            if (categories.contains(category)) {
                if (mustContainLower != null) {
                    assertThat(market.getEventType().getName().toLowerCase(), containsString(mustContainLower));
                }
            }
        }
        return counts.count(category);
    }

}

File: mns-core/src/test/java/cz/fb/manaus/core/category/CategoryServiceRealDataTest.kt
package cz.fb.manaus.core.category

import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.databind.ObjectMapper
import cz.fb.manaus.core.MarketCategories
import cz.fb.manaus.core.category.categorizer.COUNTRY_PREFIX
import cz.fb.manaus.core.category.categorizer.SPORT_PREFIX
import cz.fb.manaus.core.model.Market
import cz.fb.manaus.core.test.AbstractTestCase5
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.core.io.ResourceLoader
import java.time.Instant
import java.time.temporal.ChronoUnit
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class CategoryServiceRealDataTest : AbstractTestCase5() {
    private lateinit var markets: List<Market>

    @Autowired
    private lateinit var resourceLoader: ResourceLoader

    @Autowired
    private lateinit var objectMapper: ObjectMapper


    @BeforeEach
    fun setUp() {
        val resource = resourceLoader.getResource("classpath:cz/fb/manaus/core/service/markets.json")
        markets = objectMapper.readValue(resource.inputStream, object : TypeReference<List<Market>>() {})
        markets = markets.map { it.copy(event = it.event.copy(openDate = Instant.now().plus(5, ChronoUnit.HOURS))) }
    }

    @Autowired
    private lateinit var categoryService: CategoryService

    @Test
    fun testDisjunctiveCategories() {
        for (market in markets) {
            val marketCategories = categoryService.getMarketCategories(market, false)
            val intersection = marketCategories intersect DISJUNCTIVE_CATEGORIES
            assertEquals(1, intersection.size, market.toString())
        }
    }

    @Test
    fun testSoccer() {
        checkCategory(Category.MARKET_PREFIX + SPORT_PREFIX + MarketCategories.SOCCER, "soccer")
    }

    @Test
    fun testBasket() {
        checkCategory(Category.MARKET_PREFIX + SPORT_PREFIX + MarketCategories.BASKETBALL, "basket")
    }

    @Test
    fun testTennis() {
        checkCategory(Category.MARKET_PREFIX + SPORT_PREFIX + MarketCategories.TENNIS, "tennis")
    }

    @Test
    fun testAmericanFootball() {
        checkCategory(Category.MARKET_PREFIX + SPORT_PREFIX + MarketCategories.AMERICAN_FOOTBALL, "american")
    }

    @Test
    fun testIceHockey() {
        checkCategory(Category.MARKET_PREFIX + SPORT_PREFIX + MarketCategories.ICE_HOCKEY, "hockey")
    }

    @Test
    fun testGreyhounds() {
        checkCategory(Category.MARKET_PREFIX + SPORT_PREFIX + MarketCategories.GREY_HOUNDS, "hound")
    }

    @Test
    fun testVolleyball() {
        checkCategory(Category.MARKET_PREFIX + SPORT_PREFIX + MarketCategories.VOLLEYBALL, "volley")
    }

    @Test
    fun testGolf() {
        checkCategory(Category.MARKET_PREFIX + SPORT_PREFIX + MarketCategories.GOLF, "golf")
    }

    @Test
    fun testRunnerCount() {
        checkCategory(Category.MARKET_PREFIX + "runnerCount_" + "4", null)
    }

    @Test
    fun testUkraine() {
        checkCategory(Category.MARKET_PREFIX + COUNTRY_PREFIX + "ua", null)
    }

    @Test
    fun testFinancialBets() {
        checkCategory(Category.MARKET_PREFIX + SPORT_PREFIX + MarketCategories.FINANCIAL, "financial")
    }

    private fun checkCategory(category: String, mustContainLower: String?) {
        val size = markets.size
        val count = getCategoryCount(category, mustContainLower)
        assertTrue { count in 1 until size }
    }

    private fun getCategoryCount(category: String, mustContainLower: String?): Int {
        val counts = mutableMapOf<String, Int>().withDefault { 0 }
        for (market in markets) {
            val categories = categoryService.getMarketCategories(market, false)
            categories.forEach { counts[it] = counts.getValue(it) + 1 }
            if (category in categories) {
                if (mustContainLower != null) {
                    assertTrue { mustContainLower in market.eventType.name.lowercase() }
                }
            }
        }
        return counts[category] ?: error("no such category")
    }

    companion object {
        val DISJUNCTIVE_CATEGORIES = setOf(
            Category.MARKET_PREFIX + SPORT_PREFIX + MarketCategories.BASKETBALL,
            Category.MARKET_PREFIX + SPORT_PREFIX + MarketCategories.VOLLEYBALL,
            Category.MARKET_PREFIX + SPORT_PREFIX + MarketCategories.BASEBALL,
            Category.MARKET_PREFIX + SPORT_PREFIX + MarketCategories.SOCCER,
            Category.MARKET_PREFIX + SPORT_PREFIX + MarketCategories.AMERICAN_FOOTBALL,
            Category.MARKET_PREFIX + SPORT_PREFIX + MarketCategories.GOLF,
            Category.MARKET_PREFIX + SPORT_PREFIX + MarketCategories.HANDBALL,
            Category.MARKET_PREFIX + SPORT_PREFIX + MarketCategories.GREY_HOUNDS,
            Category.MARKET_PREFIX + SPORT_PREFIX + MarketCategories.TENNIS,
            Category.MARKET_PREFIX + SPORT_PREFIX + MarketCategories.FINANCIAL,
            Category.MARKET_PREFIX + SPORT_PREFIX + MarketCategories.SNOOKER,
            Category.MARKET_PREFIX + SPORT_PREFIX + MarketCategories.CRICKET,
            Category.MARKET_PREFIX + SPORT_PREFIX + MarketCategories.RUGBY,
            Category.MARKET_PREFIX + SPORT_PREFIX + MarketCategories.ICE_HOCKEY,
            Category.MARKET_PREFIX + SPORT_PREFIX + MarketCategories.MOTOR_SPORT,
            Category.MARKET_PREFIX + SPORT_PREFIX + MarketCategories.HORSES
        )
    }
}
